File: oxidize-python/oxidize_python/core/model/generation.py
"""Token generation streams mirroring oxidize-golang/core/model/generation.go."""

from __future__ import annotations

import math
import threading
from dataclasses import dataclass, field
from enum import Enum, auto

from oxidize_python.core.model.model import Logits, Model, Session, Token
from oxidize_python.core.model.sampling import (
    SamplingConfig,
    default_sampling_config,
    sample,
    speculative_decode,
)


class GenerationError(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(f"generation: {message}")

    def is_model_error(self) -> bool:
        return False


ERR_GENERATION_FINISHED = GenerationError("generation finished")


def is_finished(err: BaseException | None) -> bool:
    return err is not None and err is ERR_GENERATION_FINISHED


@dataclass
class GenerationConfig:
    max_new_tokens: int = 128
    stop_token: Token = 2
    stop_sequences: list[list[Token]] | None = None
    prefill_batch_size: int = 1
    sampling: SamplingConfig = field(default_factory=default_sampling_config)
    suppressed_tokens: list[Token] | None = None


def default_generation_config() -> GenerationConfig:
    return GenerationConfig(
        max_new_tokens=128,
        stop_token=2,
        stop_sequences=None,
        prefill_batch_size=1,
        sampling=default_sampling_config(),
        suppressed_tokens=None,
    )


@dataclass
class SpeculativeGenerationConfig:
    generation: GenerationConfig = field(default_factory=default_generation_config)
    draft_tokens_per_step: int = 4


def default_speculative_generation_config() -> SpeculativeGenerationConfig:
    return SpeculativeGenerationConfig(
        generation=default_generation_config(),
        draft_tokens_per_step=4,
    )


class _GenerationState(Enum):
    PREFILL = auto()
    DECODE = auto()
    DONE = auto()


class GenerationStream:
    def __init__(self, model: Model, session: Session, config: GenerationConfig) -> None:
        self._mu = threading.Lock()
        self._model = model
        self._session = session
        self._config = config
        self._state = _GenerationState.PREFILL
        self._done = False
        self._prompt: list[Token] = []
        self._generated = 0
        self._last_token: Token | None = None
        self._max_stop = max((len(s) for s in (config.stop_sequences or [])), default=0)
        self._recent: list[Token] = []

    def seed(self, prompt: list[Token]) -> None:
        with self._mu:
            if not self._prompt and prompt:
                self._prompt = list(prompt)

    def _forward_prefill(self) -> Logits:
        batch = self._config.prefill_batch_size
        if batch <= 0:
            batch = 1
        logits: Logits = []
        for i in range(0, len(self._prompt), batch):
            chunk = self._prompt[i : i + batch]
            logits = self._model.forward(chunk, self._session)
        return logits

    def _suppress_tokens(self, logits: Logits) -> None:
        for tok in self._config.suppressed_tokens or []:
            if int(tok) < len(logits):
                logits[int(tok)] = -math.inf

    def _finish_after_token(self, tok: Token) -> bool:
        if tok == self._config.stop_token:
            return True
        if self._max_stop == 0:
            return False
        self._recent.append(tok)
        if len(self._recent) > self._max_stop:
            self._recent = self._recent[-self._max_stop :]
        for seq in self._config.stop_sequences or []:
            if not seq or len(self._recent) < len(seq):
                continue
            if self._recent[-len(seq) :] == seq:
                return True
        return False

    def next(self) -> tuple[Token, bool, GenerationError | None]:
        with self._mu:
            if self._done or self._state is _GenerationState.DONE:
                return 0, True, None
            if self._config.max_new_tokens > 0 and self._generated >= self._config.max_new_tokens:
                self._done = True
                self._state = _GenerationState.DONE
                return 0, True, None
            try:
                if self._state is _GenerationState.PREFILL:
                    self._state = _GenerationState.DECODE
                    logits = self._forward_prefill()
                else:
                    if self._last_token is None:
                        self._done = True
                        self._state = _GenerationState.DONE
                        return 0, True, None
                    logits = self._model.forward([self._last_token], self._session)
            except Exception as err:
                return 0, True, GenerationError(str(err))
            if not logits:
                return 0, True, GenerationError("empty logits")
            self._suppress_tokens(logits)
            try:
                tok = sample(logits, self._config.sampling, None)
            except Exception as err:
                return 0, True, GenerationError(str(err))
            self._generated += 1
            self._last_token = tok
            if self._finish_after_token(tok):
                self._done = True
                self._state = _GenerationState.DONE
                return tok, True, None
            return tok, False, None


class SpeculativeGenerationStream:
    def __init__(
        self,
        draft: Model,
        target: Model,
        session: Session,
        config: SpeculativeGenerationConfig,
    ) -> None:
        self._mu = threading.Lock()
        self._draft = draft
        self._target = target
        self._session = session
        self._config = config
        self._done = False
        self._tokens: list[Token] = []

    def seed(self, prompt: list[Token]) -> None:
        with self._mu:
            if not self._tokens:
                self._tokens = list(prompt)

    def next(self) -> tuple[Token, bool, GenerationError | None]:
        with self._mu:
            if self._done:
                return 0, True, None
            try:
                tokens, accepted = speculative_decode(
                    self._draft,
                    self._target,
                    self._config.draft_tokens_per_step,
                    self._session,
                )
            except Exception as err:
                return 0, True, GenerationError(str(err))
            if accepted == 0:
                self._done = True
                return 0, True, None
            return tokens[accepted - 1], False, None

File: oxidize-golang/core/model/generation.go
package model

import (
	"context"
	"errors"
	"sync"
)

// GenerationConfig mirrors GenerationConfig.
type GenerationConfig struct {
	MaxNewTokens     int
	StopToken        Token
	StopSequences    [][]Token
	PrefillBatchSize int
	Sampling         SamplingConfig
	SuppressedTokens []Token
}

// DefaultGenerationConfig returns sensible defaults.
func DefaultGenerationConfig() GenerationConfig {
	return GenerationConfig{
		MaxNewTokens:     128,
		StopToken:        2,
		StopSequences:    nil,
		PrefillBatchSize: 1,
		Sampling:         DefaultSamplingConfig(),
		SuppressedTokens: nil,
	}
}

// GenerationError mirrors GenerationError.
type GenerationError struct{ Message string }

func (e *GenerationError) Error() string { return "generation: " + e.Message }

// IsModelError reports whether the error wraps a model error.
func (e *GenerationError) IsModelError() bool { return false }

// SpeculativeGenerationConfig mirrors SpeculativeGenerationConfig.
type SpeculativeGenerationConfig struct {
	Generation         GenerationConfig
	DraftTokensPerStep int
}

// DefaultSpeculativeGenerationConfig returns sensible defaults.
func DefaultSpeculativeGenerationConfig() SpeculativeGenerationConfig {
	return SpeculativeGenerationConfig{
		Generation:         DefaultGenerationConfig(),
		DraftTokensPerStep: 4,
	}
}

type generationState int

const (
	generationPrefill generationState = iota
	generationDecode
	generationDone
)

// GenerationStream is a small in-process representation of the async stream
// that the Rust crate returns from `Model::generate`. It blocks on the
// supplied model + session for each token.
type GenerationStream struct {
	mu      sync.Mutex
	model   Model
	session *Session
	config  GenerationConfig
	state   generationState
	done    bool
	prompt  []Token
	generated int
	lastToken Token
	hasLast   bool
	recent    []Token
	maxStop   int
}

// NewGenerationStream constructs a stream bound to the given model and
// session.
func NewGenerationStream(model Model, session *Session, config GenerationConfig) *GenerationStream {
	maxStop := 0
	for _, seq := range config.StopSequences {
		if len(seq) > maxStop {
			maxStop = len(seq)
		}
	}
	return &GenerationStream{
		model:   model,
		session: session,
		config:  config,
		state:   generationPrefill,
		recent:  make([]Token, 0, maxStop),
		maxStop: maxStop,
	}
}

// Seed prepends tokens to the stream's internal state.
func (s *GenerationStream) Seed(prompt []Token) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.prompt) == 0 && len(prompt) > 0 {
		s.prompt = append(s.prompt, prompt...)
	}
}

func (s *GenerationStream) forwardPrefill() (Logits, error) {
	batch := s.config.PrefillBatchSize
	if batch <= 0 {
		batch = 1
	}
	var logits Logits
	for i := 0; i < len(s.prompt); i += batch {
		end := i + batch
		if end > len(s.prompt) {
			end = len(s.prompt)
		}
		chunk := s.prompt[i:end]
		out, err := s.model.Forward(chunk, s.session)
		if err != nil {
			return nil, err
		}
		logits = out
	}
	return logits, nil
}

func (s *GenerationStream) suppressTokens(logits Logits) {
	for _, tok := range s.config.SuppressedTokens {
		if int(tok) < len(logits) {
			logits[tok] = negInfF32
		}
	}
}

func (s *GenerationStream) finishAfterToken(tok Token) bool {
	if tok == s.config.StopToken {
		return true
	}
	if s.maxStop == 0 {
		return false
	}
	s.recent = append(s.recent, tok)
	if len(s.recent) > s.maxStop {
		s.recent = s.recent[len(s.recent)-s.maxStop:]
	}
	for _, seq := range s.config.StopSequences {
		if len(seq) == 0 || len(s.recent) < len(seq) {
			continue
		}
		ok := true
		for i, x := range seq {
			if s.recent[len(s.recent)-len(seq)+i] != x {
				ok = false
				break
			}
		}
		if ok {
			return true
		}
	}
	return false
}

// Next produces the next token. When generation is finished, returns
// (token=0, done=true, nil). Errors are wrapped in GenerationError.
func (s *GenerationStream) Next(ctx context.Context) (Token, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.done || s.state == generationDone {
		return 0, true, nil
	}
	if s.config.MaxNewTokens > 0 && s.generated >= s.config.MaxNewTokens {
		s.done = true
		s.state = generationDone
		return 0, true, nil
	}
	if err := ctx.Err(); err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}

	var logits Logits
	var err error
	switch s.state {
	case generationPrefill:
		s.state = generationDecode
		logits, err = s.forwardPrefill()
	case generationDecode:
		if !s.hasLast {
			s.done = true
			s.state = generationDone
			return 0, true, nil
		}
		logits, err = s.model.Forward([]Token{s.lastToken}, s.session)
	default:
		return 0, true, nil
	}
	if err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	if len(logits) == 0 {
		return 0, true, &GenerationError{Message: "empty logits"}
	}
	s.suppressTokens(logits)
	tok, err := Sample(logits, s.config.Sampling, nil)
	if err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	s.generated++
	s.lastToken = tok
	s.hasLast = true
	if s.finishAfterToken(tok) {
		s.done = true
		s.state = generationDone
		return tok, true, nil
	}
	return tok, false, nil
}

// SpeculativeGenerationStream mirrors SpeculativeGenerationStream<'a, T: Model>.
type SpeculativeGenerationStream struct {
	mu      sync.Mutex
	draft   Model
	target  Model
	session *Session
	config  SpeculativeGenerationConfig
	done    bool
}

// NewSpeculativeGenerationStream constructs a speculative stream using a
// draft + target model.
func NewSpeculativeGenerationStream(draft, target Model, session *Session, config SpeculativeGenerationConfig) *SpeculativeGenerationStream {
	return &SpeculativeGenerationStream{draft: draft, target: target, session: session, config: config}
}

// Seed runs a prefill on the target model for the prompt tokens.
func (s *SpeculativeGenerationStream) Seed(prompt []Token) {
	if len(prompt) == 0 {
		return
	}
	_, _ = s.target.Forward(prompt, s.session)
}

// Next returns the next accepted draft token plus a done flag.
func (s *SpeculativeGenerationStream) Next(ctx context.Context) (Token, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.done {
		return 0, true, nil
	}
	if err := ctx.Err(); err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	tokens, accepted, err := SpeculativeDecode(s.draft, s.target, s.config.DraftTokensPerStep, s.session)
	if err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	if accepted == 0 {
		s.done = true
		return 0, true, nil
	}
	return tokens[accepted-1], false, nil
}

// errGenerationFinished is returned when a stream has been fully consumed.
var errGenerationFinished = errors.New("generation finished")

// IsFinished reports whether err is errGenerationFinished.
func IsFinished(err error) bool { return errors.Is(err, errGenerationFinished) }
